tools/mkldscript.c

```diff
@@ -9,6 +9,10 @@
 #include "spec.h"
 #include "util.h"
 
+// Note: *SECTION ALIGNMENT* Object files built with a compiler such as GCC can, by default, use narrower alignment 
+// for sections compared to IDO's 0x10. To properly generate relocations relative to section starts, sections currently 
+// need to be aligned to 0x10, because otherwise the linker inserts padding in the wrong place.
```

Comment:
when this says "currently", is it a Fado limitation or a link-time-related limitation?

Reply:
It is, as this says, because the linker puts the padding in the wrong place, namely after the section ends. Fado would be able to handle it if the section sizes were computed including padding, but they aren't (and you can't expect Fado to know what the linker script feels like doing). Indeed, it has an experimental option to do use the alignment the elf files specify.

Reply:
## The problem

The alignment problem is related to the `_SEGMENTNAMESegmentCAT` symbols generated by the linker.
For example, when compiling with `mips64-gcc` and not applying the `ALIGN(0x10)` patch in the linker script, you get this in the map:
```
 build/src/overlays/actors/ovl_Object_Kankyo/z_object_kankyo.o(.data)
 .data          0x0000000080b72e6c       0x76 build/src/overlays/actors/ovl_Object_Kankyo/z_object_kankyo.o
 build/src/overlays/actors/ovl_Object_Kankyo/ovl_Object_Kankyo_reloc.o(.data)
 .data          0x0000000080b72ee2        0x0 build/src/overlays/actors/ovl_Object_Kankyo/ovl_Object_Kankyo_reloc.o
                0x0000000080b72ee2                _ovl_Object_KankyoSegmentDataEnd = .
                0x0000000000000076                _ovl_Object_KankyoSegmentDataSize = ABSOLUTE ((_ovl_Object_KankyoSegmentDataEnd - _ovl_Object_KankyoSegmentDataStart))
                0x0000000080b72ee2                _ovl_Object_KankyoSegmentRoDataStart = .
 build/src/overlays/actors/ovl_Object_Kankyo/z_object_kankyo.o(.rodata)
 *fill*         0x0000000080b72ee2        0x6 
 .rodata        0x0000000080b72ee8      0x1b8 build/src/overlays/actors/ovl_Object_Kankyo/z_object_kankyo.o
                0x0000000080b72ee8                Object_Kankyo_InitVars
 build/src/overlays/actors/ovl_Object_Kankyo/z_object_kankyo.o(.rodata.str1.4)
 build/src/overlays/actors/ovl_Object_Kankyo/z_object_kankyo.o(.rodata.cst4)

```
This causes `_ovl_Object_KankyoSegmentRoDataStart` to not be properly aligned (0x0000000080b72ee2). This is a problem because `relocation.c` expects word alignment on sections (it does pointer arithmetic and just dereferences the pointers without checking for them to be correctly aligned).
Also, that number is lying, because, as you can see, the first rodata file is at VRAM `0x0000000080b72ee8`, this causes the relocs to be "wrong" compared to the linked program, and `relocation.c` will use the wrong offsets for its calculations.

## Why this happens?

First, look at the size of the .data section of z_object_kankyo
```
 build/src/overlays/actors/ovl_Object_Kankyo/z_object_kankyo.o(.data)
 .data          0x0000000080b72e6c       0x76 build/src/overlays/actors/ovl_Object_Kankyo/z_object_kankyo.o
 ```
It has size 0x76, which is not word aligned. Weird, right?, let's take a look at `readelf` (`readelf -a  build/src/overlays/actors/ovl_Object_Kankyo/z_object_kankyo.o`)

```
Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  // other sections...
  [ 3] .data             PROGBITS        00000000 002710 000076 00  WA  0   0  4
  // other sections...
  [11] .rodata           PROGBITS        00000000 002a18 0001b8 00   A  0   0  8
```

The object itself says the .data section has size 0x76, with a 4 byte alignment (last column). The starting address for the data is already 4-aligned (`0x0000000080b72e6c`), so the linker doesn't have to adjust anything. The address is just incremented to account for this size (ending up not word aligned).

Then, according to our linker script, the linker has to produce the `SegmentDataEnd` and `SegmentRoDataStart` symbols, just taking the "current" processed address.

After that, the rodata is processed. As you can see in the readelf output, it requires 8-byte aligment, and since the current address (`0x0000000080b72ee2`) is not 8-byte aligned, then it adds a `*fill*` of 6 bytes. This `*fill*` introduce an empty space between the `SegmentRoDataStart` and the real rodata start.

## Can it be fixed in the side of Fado?

Not really. Fado currently forces each section to be 0x10 aligned, but as Elliptic said, there's an experimental option to account for the alignment declared in the .o file which is being processed.

But, Fado relies on the `SegmentSECTIONSize` symbols to be correct, which are produced by the linker, so there isn't much Fado can do it this regard.

## Can it be fixed in the linker script (mkldscript)?

Maybe? I'm no linker script expert, but I think it would be hard to patch in the mkldscript.c itself. Mainly because the alignment is accounted by the linker _after_ the `SegmentSECTIONStart` symbol has been emitted, and I don't see any way to fix the alignment of the section, then emitting the symbol and then writing the sections themselves

Reply:
:sunglasses: thanks for the explanation!

So in the comment in mkldscript, "use narrower alignment ... for sections" refers to alignment of the section sizes?

So for example the data section of `z_object_kankyo.o` in IDO is always padded to a 0x10-alignment? which means the linker has no further alignment to do since an align size added to an aligned address is an aligned address
And with gcc the data section is of unaligned/unpadded size so the linker has to introduce alignment on the resulting address

If that's right the mkldscript comment on the aligns may be fleshed out a bit about aligning size vs aligning addresses or something, it's too fresh for me to word something

I will also say, it looks more worthwhile to indeed just align everything, rather than try to get stuff to work with the linker-made padding

Reply:
> So in the comment in mkldscript, "use narrower alignment ... for sections" refers to alignment of the section sizes?

Yeah, pretty much.

> So for example the data section of z_object_kankyo.o in IDO is always padded to a 0x10-alignment?

Yes, IDO always aligns each section to 0x10. Actually, that's the basic way to find file splits in multi-file segments, like `code`, `boot` and some overlays.

> which means the linker has no further alignment to do since an align size added to an aligned address is an aligned address
> And with gcc the data section is of unaligned/unpadded size so the linker has to introduce alignment on the resulting address

Exactly

> I will also say, it looks more worthwhile to indeed just align everything, rather than try to get stuff to work with the linker-made padding

Yeah, that what every `ALIGN(0x10)` in every section is doing

Reply:
So to summarize your detailed explanation into the comment, what about

Note: \*SECTION ALIGNMENT\* Object files built with a compiler such as GCC can, by default, use narrower alignment for sections size, compared to IDO padding sections to a 0x10-aligned size.
To properly generate relocations relative to section starts, sections currently need to be aligned explicitly (to 0x10 currently, a narrower alignment might work), otherwise the linker does implicit alignment and inserts padding between the address indicated by section start symbols (such as `*SegmentRoDataStart`) and the actual aligned start of the section.
With IDO, the padding of sections to an aligned size makes the section start at aligned addresses out of the box, so the explicit alignment has no further effect.

Reply:
Yeah, it sounds good to me. :+1: